### PR TITLE
support generic variant setter

### LINKED ISSUE #50355: Calling setter on variant accessors with function using type parameter errors
# Bug Report

### 🔎 Search Terms

- Variant accessors / generics / functions

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (4.35 -> nightly), and I reviewed the FAQ for entries about **variant accessors**.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.7.4#code/MYGwhgzhAEAqCmEAuAeAygPmgbwFDQOmAHsA7ZAJwFdgliKAKAShwF9d9CJ4loA3MCCrwGAofABc0SgEtSAc2gAfaAwYyk8ALZS0LALxZZCpi2ztOBeT36DhzKccV5Cr6BR5UKpaKSogQAEIAbktodlcwgHoo6AAleAAHcGA5RSQAC3hoayRNCmkaDOhMsF5M7Nz8qO48+AKkAE9ExGgtMuAMgBp3bWI+Voroeop6CWjY3NtxB2kkCjTlVXVNHWg9aEM5hZMcCbcPJC8fPwCQ-YjCfYB5Ch6PZLBUhRKs6R58ws7XsugwIjIwEO2SaLWgMhgVG4ABN7n0BjAhiMxvtatN7GJhI55osVGoNNopNhoIlRi0KE1sTtFKwDEYcSYzOwLCRyLx-vpfPAAO5wRCoFyEUnEcmU7ZpXCsDDMUK4MAAOkx2U5K20dPBq3lwtFjWgMWgAAUyfUmtAAOTak2NM3QaHEVqkYi8eAADwhvDIJWa2TNaDNQA)

### 💻 Code

```ts
class Test<S> {
    constructor() {}

    set value(value: string | ((item: S) => string)) {}

    get value(): string {
        return null!;
    }
    
    // -- Replacing the getter such that the getter/setter types match, removes the error:
    // get value(): string | ((item: S) => string) {
    //     return null!;
    // }
    
    // -- Or, replacing the setter such that a concrete type is used, removes the error:
    // set value(value: string | ((item: { property: string }) => string)) {}
}

const a = new Test<{
    property: string
}>();

a.value = (item) => item.property // Property 'property' does not exist on type 'S'
```

### 🙁 Actual behavior

Using variant accessors, with the setter taking a function with a class generic type as an argument, compilation fails with the
error `Property 'property' does not exist on type 'S'` (marked in code example above). This error does not occur if the type of the getter matches the setter, if the getter is removed, or a concrete type is specified instead of a type parameter in the setter.

### 🙂 Expected behavior

No error with support for inferring the generic type `S` when assigning with the setter.

### PR DESCRIPTION
**Pull request by Zzzen**
Fixes #50355

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -18760,7 +18760,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function instantiateSymbol(symbol: Symbol, mapper: TypeMapper): Symbol {
         const links = getSymbolLinks(symbol);
-        if (links.type && !couldContainTypeVariables(links.type)) {
+        if (links.type && !couldContainTypeVariables(links.type) && !(getDeclarationOfKind(symbol, SyntaxKind.SetAccessor) && couldContainTypeVariables(getWriteTypeOfSymbol(symbol)))) {
```
Comment (jakebailey) [inline src/compiler/checker.ts:18763]: I get the gut feeling that this check is in the wrong place, but I'm not totally sure... This block to me seems like a fast path, e.g. "if there's no type variables then we don't need to do anything because instantiation is a no-op", but this new logic feels tangential.

Indeed if you remove this block entirely, all tests pass (except one tooltip test which I even think is wrong), which would imply that this isn't quite the right place.

But, I'm not totally sure without debugging to see where this is now firing.

Comment (jakebailey) [reply src/compiler/checker.ts:18763]: I did a bit of debugging to understand the problem better.

My understanding is that this block is indeed an optimization, but it has a bug; if the symbol has a setter, then the write type also should be checked for the potential of type variables. Hence why removing this block altogether fixes the issue; it's best effort and it's okay if you don't do it. But it's not okay if you do it and are wrong!

To that end, I don't think that `getWriteTypeOfSymbol` is a good idea here because that may accidentally pull on things; there's a reason the original check wasn't using `getTypeOfSymbol`, for example. But, I do think that there's a version of this that would be more congruent to the existing check which also fixes the bug:

```ts
// If the type of the symbol is already resolved, and if that type could not possibly
// be affected by instantiation, simply return the symbol itself.
if (links.type && !couldContainTypeVariables(links.type)) {
    if (!(symbol.flags & SymbolFlags.SetAccessor)) {
        return symbol;
    }
    // If we're a setter, check writeType.
    if (links.writeType && !couldContainTypeVariables(links.writeType)) {
        return symbol;
    }
}
```

So, basically just the original code, but if it's a setter, we check `writeType` too.

**Review by jakebailey — CHANGES_REQUESTED** [review]

File: src/compiler/checker.ts (hunk 1 above)
Comment (Zzzen) [reply src/compiler/checker.ts:18763]: Awesome, I didn't realize there was a "writeTypes" in the links.

Comment (jakebailey) [reply src/compiler/checker.ts:18763]: FWIW that PR is #55388 and it seems like it could actually matter for perf (slight memory, maybe one benchmark changed).

**Comment by jakebailey** [comment]
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster

**Comment by jakebailey** [comment]
sigh

@typescript-bot run dt

**Review by jakebailey — APPROVED** [review]
LGTM now, but @weswigham may be able to poke holes.
